File: javascript/book-store/book-store.js
// book-store.js
const BOOK_PRICE = 800; // Price in cents
const DISCOUNTS = {
  1: 1.0,  // 0% discount
  2: 0.95, // 5% discount
  3: 0.90, // 10% discount
  4: 0.80, // 20% discount
  5: 0.75, // 25% discount
};

export const cost = (books) => {
  if (!books || books.length === 0) {
    return 0;
  }

  // Count occurrences of each book
  const counts = {};
  for (const book of books) {
    counts[book] = (counts[book] || 0) + 1;
  }

  // Iteratively form sets of distinct books
  const setSizes = [];
  let distinctBooks = Object.keys(counts).filter(book => counts[book] > 0);

  while (distinctBooks.length > 0) {
    setSizes.push(distinctBooks.length);
    // Decrement count for books included in this set
    for (const book of distinctBooks) {
      counts[book]--;
    }
    // Update the list of distinct books available for the next set
    distinctBooks = Object.keys(counts).filter(book => counts[book] > 0);
  }

  // Apply the optimization: replace sets of 5 and 3 with two sets of 4
  // This is cheaper: (5 * 0.75 + 3 * 0.90) * 800 = (3.75 + 2.70) * 800 = 6.45 * 800
  // vs (4 * 0.80 + 4 * 0.80) * 800 = (3.20 + 3.20) * 800 = 6.40 * 800
  while (setSizes.includes(5) && setSizes.includes(3)) {
    const indexOf5 = setSizes.indexOf(5);
    setSizes.splice(indexOf5, 1); // Remove one 5
    const indexOf3 = setSizes.indexOf(3);
    setSizes.splice(indexOf3, 1); // Remove one 3
    setSizes.push(4, 4); // Add two 4s
  }

  // Calculate the total cost
  let totalCost = 0;
  for (const size of setSizes) {
    totalCost += size * BOOK_PRICE * DISCOUNTS[size];
  }

  // Return the total cost in cents (should already be integer)
  return Math.round(totalCost); // Use Math.round just in case of floating point inaccuracies
};

File: java/book-store/src/main/java/BookStore.java
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Map;
import java.util.HashMap;

class BookStore {

    private static final double BOOK_PRICE = 8.0;
    // Discounts map: group size -> discount multiplier
    private static final Map<Integer, Double> DISCOUNTS = Map.of(
            1, 1.0,   // 0%
            2, 0.95,  // 5%
            3, 0.90,  // 10%
            4, 0.80,  // 20%
            5, 0.75   // 25%
    );

    /**
     * Calculates the minimum cost for a basket of books, applying discounts
     * for sets of different books. It optimizes for the case where two groups
     * of 4 are cheaper than a group of 5 and a group of 3.
     *
     * @param books A list of integers representing the book IDs (assumed to be 1-5).
     * @return The minimum possible cost for the basket.
     */
    double calculateBasketCost(List<Integer> books) {
        if (books == null || books.isEmpty()) {
            return 0.0;
        }

        // 1. Count occurrences of each book type
        Map<Integer, Integer> counts = new HashMap<>();
        for (int book : books) {
            // We only care about the distinct books mentioned in the problem (1-5)
            // Although the test cases might use other numbers, the discount logic
            // only applies based on the *number* of distinct books per set.
            counts.put(book, counts.getOrDefault(book, 0) + 1);
        }

        // 2. Greedily form discount groups of distinct books
        List<Integer> groupSizes = new ArrayList<>();
        while (!counts.isEmpty()) {
            int currentSetSize = counts.size(); // Number of distinct books remaining
            groupSizes.add(currentSetSize);

            // Decrement the count for each distinct book type included in this set
            List<Integer> booksToRemoveFromMap = new ArrayList<>();
            for (Map.Entry<Integer, Integer> entry : counts.entrySet()) {
                int book = entry.getKey();
                int currentCount = entry.getValue();
                if (currentCount == 1) {
                    // If this was the last copy, mark it for removal from the map
                    booksToRemoveFromMap.add(book);
                } else {
                    // Otherwise, just decrement the count
                    counts.put(book, currentCount - 1);
                }
            }
            // Remove books that have no copies left
            for (int book : booksToRemoveFromMap) {
                counts.remove(book);
            }
        }

        // 3. Apply the optimization: Replace (5, 3) pairs with (4, 4) pairs
        //    as long as possible, because 2*Cost(4) < Cost(5) + Cost(3)
        while (groupSizes.contains(5) && groupSizes.contains(3)) {
            groupSizes.remove(Integer.valueOf(5)); // Remove one instance of 5
            groupSizes.remove(Integer.valueOf(3)); // Remove one instance of 3
            groupSizes.add(4);                     // Add a 4
            groupSizes.add(4);                     // Add another 4
        }

        // 4. Calculate the total cost based on the final group sizes
        double totalCost = 0.0;
        for (int size : groupSizes) {
             if (size > 0 && DISCOUNTS.containsKey(size)) {
                 // Apply discount based on the size of the group
                 totalCost += size * BOOK_PRICE * DISCOUNTS.get(size);
             } else if (size > 0) {
                 // Should theoretically only be size 1 if not in DISCOUNTS map,
                 // or potentially > 5 if input books were outside 1-5 range,
                 // but discount only applies up to 5 distinct books.
                 // Treat as individual books if size isn't in discount map.
                 totalCost += size * BOOK_PRICE;
             }
             // size == 0 is ignored
        }

        // Let's return the raw double, tests might handle precision.
        return totalCost;
    }
}